<a href="https://colab.research.google.com/github/akutayaydin/Magnimind-3/blob/main/Twitter_NLP_v7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Twitter NLP Classification Project

Goal: Create a ML model for Business Category tweets

In [5]:
import pandas as pd

In [6]:
#For Google Colab
url ='https://raw.githubusercontent.com/akutayaydin/Magnimind-3/main/tweet_data_small_filtered.csv'
df = pd.read_csv(url)

In [7]:
#df = pd.read_csv('tweet_data.csv')

In [8]:
#Select Business Category Only as Request
df = df[df['topicName'] == 'Business']

In [9]:
df.head(10)

tweetID               crDate  edInput  editor  engages  \
0  1072758722315862021  2018-12-12 07:43:04        2    5001     9618   
1  1072959907958276096  2018-12-12 21:02:31        2    5001    10388   
2  1073329348755767303  2018-12-13 21:30:32        2    5001     6734   
3  1073499538659393536  2018-12-14 08:46:49        1    5001     6914   
4  1073542862573371392  2018-12-14 11:38:58        2    5001     6269   
5  1074475798227226627  2018-12-17 01:26:07        2    5001     7701   
6  1074056878542471169  2018-12-15 21:41:29        1    5001     6284   
7  1074964829830557696  2018-12-18 09:49:21        2    5001     8678   
8  1075269960086241281  2018-12-19 06:01:50        1    5001      786   
9  1075315807666794496  2018-12-19 09:04:01        1    5001      664   

   isApproved  isEdNeed   isRT  likes  \
0       False      True   True   5771   
1       False      True   True   6977   
2       False      True   True   4467   
3        True      True   True   5402   
4       False      True   True   4610   
5       False      True  False   5730   
6        True      True   True   5207   
7       False      True   True   5206   
8        True      True  False    548   
9        True      True  False    436   

                                          photoUrl  retweets   rtUsID  \
0                                              NaN      3847   621523   
1  https://pbs.twimg.com/media/DuPrWYHW4AEXol3.jpg      3411   621523   
2                                              NaN      2267  3108351   
3                                              NaN      1512   621523   
4                                              NaN      1659   621523   
5                                              NaN      1971       -1   
6                                              NaN      1077  2735591   
7  https://pbs.twimg.com/media/DusKzteX4AASYwo.jpg      3472   621523   
8                                              NaN       238       -1   
9                                              NaN       228       -1   

                                                text topicName   usFlwrs  \
0  UK Prime Minister Theresa May will face a vote...  Business  38564612   
1  UK PM Theresa May wins confidence vote with 20...  Business  38567131   
2  The probe of the inaugural fund partly arises ...  Business     57108   
3  The week Brexit hit the brick wall :  Commons ...  Business    879928   
4  Have watched these kinds of pictures looking f...  Business    888862   
5  The world's most successful people start their...  Business   2700458   
6  “This is Pearl Harbor. The whole country, and ...  Business     12423   
7  Jose Mourinho has been sacked by Manchester Un...  Business   7573834   
8  Pursue your ideas, no matter how obvious they ...  Business   5376604   
9  You don’t have to choose between being a leade...  Business   5376647   

        usID              usName  \
0    5402612   BBC Breaking News   
1    5402612   BBC Breaking News   
2  705706292    Rebecca Ballhaus   
3   25984418       Nick Robinson   
4   61183568    Laura Kuenssberg   
5   16896485                Inc.   
6   41154223  Saikat Chakrabarti   
7  265902729           BBC Sport   
8   14800270  Harvard Biz Review   
9   14800270  Harvard Biz Review   

                                            videoUrl  
0                                                NaN  
1                                                NaN  
2                                                NaN  
3                                                NaN  
4  https://video.twimg.com/ext_tw_video/107354281...  
5                                                NaN  
6                                                NaN  
7                                                NaN  
8                                                NaN  
9                                                NaN

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30024 entries, 0 to 30023
Data columns (total 18 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   tweetID     30024 non-null  int64 
 1   crDate      30024 non-null  object
 2   edInput     30024 non-null  int64 
 3   editor      30024 non-null  int64 
 4   engages     30024 non-null  int64 
 5   isApproved  30024 non-null  bool  
 6   isEdNeed    30024 non-null  bool  
 7   isRT        30024 non-null  bool  
 8   likes       30024 non-null  int64 
 9   photoUrl    2389 non-null   object
 10  retweets    30024 non-null  int64 
 11  rtUsID      30024 non-null  int64 
 12  text        30024 non-null  object
 13  topicName   30024 non-null  object
 14  usFlwrs     30024 non-null  int64 
 15  usID        30024 non-null  int64 
 16  usName      30024 non-null  object
 17  videoUrl    7184 non-null   object
dtypes: bool(3), int64(9), object(6)
memory usage: 3.8+ MB


>Preprocessing starts here

In [11]:
df_section = df[['tweetID','text','topicName','usID','usName','edInput','crDate']].copy()

We dont need these columns.
5 unique values for edInput : (5, array([-1,  2,  1,  4,  3]))
2 unique values for isApproved: (2, array([False,  True]))
12 unique values for editor: (12,
 array([  -1, 5001, 5003, 5004, 5002, 1001, 2001, 5006, 5101, 5005, 5008,
        5007]))
42 unique values for topicname:(42,
 array(['Business', 'Animal', 'Memes', 'Nature', 'Travel', 'Interesting',
        'News', 'Fashion', 'Art', 'Personal Development', 'Motivational',
        'DIY & Craft', 'Editorial Pics', 'Funny', 'Poem', 'World Leaders',
        'Entrepreneurship', 'US News', 'Journalists', 'Flowers',
        'Cryptocurrency', 'Photography', 'Design & Architecture',
        'Fitness', 'Spirituality', 'Startup', 'Space', 'Technology',
        'Digital Media & Internet', 'Psychology', 'Science', 'Basketball',
        'Humor', 'Mindfulness', 'Medical', 'Informative', 'Love', 'Stocks',
        'Nanotech', 'Biotech', 'science', 'Random'], dtype=object))

In [12]:
#Remove duplicates from the dataframe

In [13]:
df_section = df_section.drop_duplicates(subset='text')

In [14]:
df_section.head()

tweetID                                               text  \
0  1072758722315862021  UK Prime Minister Theresa May will face a vote...   
1  1072959907958276096  UK PM Theresa May wins confidence vote with 20...   
2  1073329348755767303  The probe of the inaugural fund partly arises ...   
3  1073499538659393536  The week Brexit hit the brick wall :  Commons ...   
4  1073542862573371392  Have watched these kinds of pictures looking f...   

  topicName       usID             usName  edInput               crDate  
0  Business    5402612  BBC Breaking News        2  2018-12-12 07:43:04  
1  Business    5402612  BBC Breaking News        2  2018-12-12 21:02:31  
2  Business  705706292   Rebecca Ballhaus        2  2018-12-13 21:30:32  
3  Business   25984418      Nick Robinson        1  2018-12-14 08:46:49  
4  Business   61183568   Laura Kuenssberg        2  2018-12-14 11:38:58

>Preprocessing ends here

>Text Processing and Tokenization starts here

In [15]:
#1. Tokenize text column and save it as a list in tokens column
#2. Convert all words in the tokens column to lowercase
#3. Remove Stopwords
#4. Find all urls and parse them to a new column (tokens_url) >> TODO: Later this url(s) will be scraped to get more information about the category. 
#5. Remove Punctuations and create a new column (tokens_new)   

In [16]:
import re
from nltk.tokenize import TweetTokenizer
from nltk.corpus import stopwords

In [17]:
tknz =TweetTokenizer()

In [18]:
df_section['tokens'] =''

In [19]:
df_section['tokens'] = df['text'].apply(tknz.tokenize)

In [20]:
df_section.head()

tweetID                                               text  \
0  1072758722315862021  UK Prime Minister Theresa May will face a vote...   
1  1072959907958276096  UK PM Theresa May wins confidence vote with 20...   
2  1073329348755767303  The probe of the inaugural fund partly arises ...   
3  1073499538659393536  The week Brexit hit the brick wall :  Commons ...   
4  1073542862573371392  Have watched these kinds of pictures looking f...   

  topicName       usID             usName  edInput               crDate  \
0  Business    5402612  BBC Breaking News        2  2018-12-12 07:43:04   
1  Business    5402612  BBC Breaking News        2  2018-12-12 21:02:31   
2  Business  705706292   Rebecca Ballhaus        2  2018-12-13 21:30:32   
3  Business   25984418      Nick Robinson        1  2018-12-14 08:46:49   
4  Business   61183568   Laura Kuenssberg        2  2018-12-14 11:38:58   

                                              tokens  
0  [UK, Prime, Minister, Theresa, May, will, face...  
1  [UK, PM, Theresa, May, wins, confidence, vote,...  
2  [The, probe, of, the, inaugural, fund, partly,...  
3  [The, week, Brexit, hit, the, brick, wall, :, ...  
4  [Have, watched, these, kinds, of, pictures, lo...

In [21]:
df_section.shape

(30021, 8)

In [22]:
#Convert the list to lowercase
def convert_string_lower(strings):
    newList = []
    for string in strings:
        string = string.replace('\'','') # don't loose words with these character
        string = string.replace('-','') # don't loose words with these character
        newList.append(string.lower())
    return newList

df_section['tokens'] = list(map(convert_string_lower,df_section['tokens'])) 

In [23]:
df_section.head(5)

tweetID                                               text  \
0  1072758722315862021  UK Prime Minister Theresa May will face a vote...   
1  1072959907958276096  UK PM Theresa May wins confidence vote with 20...   
2  1073329348755767303  The probe of the inaugural fund partly arises ...   
3  1073499538659393536  The week Brexit hit the brick wall :  Commons ...   
4  1073542862573371392  Have watched these kinds of pictures looking f...   

  topicName       usID             usName  edInput               crDate  \
0  Business    5402612  BBC Breaking News        2  2018-12-12 07:43:04   
1  Business    5402612  BBC Breaking News        2  2018-12-12 21:02:31   
2  Business  705706292   Rebecca Ballhaus        2  2018-12-13 21:30:32   
3  Business   25984418      Nick Robinson        1  2018-12-14 08:46:49   
4  Business   61183568   Laura Kuenssberg        2  2018-12-14 11:38:58   

                                              tokens  
0  [uk, prime, minister, theresa, may, will, face...  
1  [uk, pm, theresa, may, wins, confidence, vote,...  
2  [the, probe, of, the, inaugural, fund, partly,...  
3  [the, week, brexit, hit, the, brick, wall, :, ...  
4  [have, watched, these, kinds, of, pictures, lo...

In [24]:
import nltk
from nltk.corpus import stopwords

In [25]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [26]:
#remove stopwords      
nltk_words = stopwords.words('english') 

def remove_stopwords(string):
    output = [w for w in string if w not in nltk_words]
    return output

df_section['tokens'] = df_section['tokens'].apply(remove_stopwords)

In [27]:
df_section.head(10)

tweetID                                               text  \
0  1072758722315862021  UK Prime Minister Theresa May will face a vote...   
1  1072959907958276096  UK PM Theresa May wins confidence vote with 20...   
2  1073329348755767303  The probe of the inaugural fund partly arises ...   
3  1073499538659393536  The week Brexit hit the brick wall :  Commons ...   
4  1073542862573371392  Have watched these kinds of pictures looking f...   
5  1074475798227226627  The world's most successful people start their...   
6  1074056878542471169  “This is Pearl Harbor. The whole country, and ...   
7  1074964829830557696  Jose Mourinho has been sacked by Manchester Un...   
8  1075269960086241281  Pursue your ideas, no matter how obvious they ...   
9  1075315807666794496  You don’t have to choose between being a leade...   

  topicName       usID              usName  edInput               crDate  \
0  Business    5402612   BBC Breaking News        2  2018-12-12 07:43:04   
1  Business    5402612   BBC Breaking News        2  2018-12-12 21:02:31   
2  Business  705706292    Rebecca Ballhaus        2  2018-12-13 21:30:32   
3  Business   25984418       Nick Robinson        1  2018-12-14 08:46:49   
4  Business   61183568    Laura Kuenssberg        2  2018-12-14 11:38:58   
5  Business   16896485                Inc.        2  2018-12-17 01:26:07   
6  Business   41154223  Saikat Chakrabarti        1  2018-12-15 21:41:29   
7  Business  265902729           BBC Sport        2  2018-12-18 09:49:21   
8  Business   14800270  Harvard Biz Review        1  2018-12-19 06:01:50   
9  Business   14800270  Harvard Biz Review        1  2018-12-19 09:04:01   

                                              tokens  
0  [uk, prime, minister, theresa, may, face, vote...  
1  [uk, pm, theresa, may, wins, confidence, vote,...  
2  [probe, inaugural, fund, partly, arises, mater...  
3  [week, brexit, hit, brick, wall, :, commons, ’...  
4  [watched, kinds, pictures, looking, body, lang...  
5  [worlds, successful, people, start, day, 4, .,...  
6  [“, pearl, harbor, ., whole, country, ,, whole...  
7  [jose, mourinho, sacked, manchester, united, ....  
8  [pursue, ideas, ,, matter, obvious, seem, http...  
9  [’, choose, leader, life, https://t.co/sc9fi5v...

In [28]:
#Check if any stop words exists
df_section[['the' in x for x in df_section['tokens']]]

Empty DataFrame
Columns: [tweetID, text, topicName, usID, usName, edInput, crDate, tokens]
Index: []

In [29]:
#find urls and create a new column for them
#TODO: Scrape this url to get more info     
url_pattern = re.compile(r"([http#]\W+)")

def find_url(string):
    output = [w for w in string if url_pattern.search(w)]
    return output

df_section['tokens_url'] = df_section['tokens'].apply(find_url)


In [30]:
df_section.head(10)

tweetID                                               text  \
0  1072758722315862021  UK Prime Minister Theresa May will face a vote...   
1  1072959907958276096  UK PM Theresa May wins confidence vote with 20...   
2  1073329348755767303  The probe of the inaugural fund partly arises ...   
3  1073499538659393536  The week Brexit hit the brick wall :  Commons ...   
4  1073542862573371392  Have watched these kinds of pictures looking f...   
5  1074475798227226627  The world's most successful people start their...   
6  1074056878542471169  “This is Pearl Harbor. The whole country, and ...   
7  1074964829830557696  Jose Mourinho has been sacked by Manchester Un...   
8  1075269960086241281  Pursue your ideas, no matter how obvious they ...   
9  1075315807666794496  You don’t have to choose between being a leade...   

  topicName       usID              usName  edInput               crDate  \
0  Business    5402612   BBC Breaking News        2  2018-12-12 07:43:04   
1  Business    5402612   BBC Breaking News        2  2018-12-12 21:02:31   
2  Business  705706292    Rebecca Ballhaus        2  2018-12-13 21:30:32   
3  Business   25984418       Nick Robinson        1  2018-12-14 08:46:49   
4  Business   61183568    Laura Kuenssberg        2  2018-12-14 11:38:58   
5  Business   16896485                Inc.        2  2018-12-17 01:26:07   
6  Business   41154223  Saikat Chakrabarti        1  2018-12-15 21:41:29   
7  Business  265902729           BBC Sport        2  2018-12-18 09:49:21   
8  Business   14800270  Harvard Biz Review        1  2018-12-19 06:01:50   
9  Business   14800270  Harvard Biz Review        1  2018-12-19 09:04:01   

                                              tokens  \
0  [uk, prime, minister, theresa, may, face, vote...   
1  [uk, pm, theresa, may, wins, confidence, vote,...   
2  [probe, inaugural, fund, partly, arises, mater...   
3  [week, brexit, hit, brick, wall, :, commons, ’...   
4  [watched, kinds, pictures, looking, body, lang...   
5  [worlds, successful, people, start, day, 4, .,...   
6  [“, pearl, harbor, ., whole, country, ,, whole...   
7  [jose, mourinho, sacked, manchester, united, ....   
8  [pursue, ideas, ,, matter, obvious, seem, http...   
9  [’, choose, leader, life, https://t.co/sc9fi5v...   

                                          tokens_url  
0                          [https://t.co/wzta1v9scu]  
1  [https://t.co/9vw2goqodc, https://t.co/bafuusz...  
2                          [https://t.co/wvercqwnvg]  
3                                                 []  
4                          [https://t.co/1tg4h62z7l]  
5                          [https://t.co/sef0u4xwp8]  
6                          [https://t.co/vposikgvim]  
7  [https://t.co/xl8mju4w1u, https://t.co/zdrd3vu...  
8                          [https://t.co/irqq39x9h2]  
9                          [https://t.co/sc9fi5v8zs]

In [31]:
df_section['tokens_new'] = ''

In [32]:
#remove punctuations

def no_punc(string):
  no_punc = [w for w in string if w.isalpha()]
  return no_punc

df_section['tokens_new'] = df_section['tokens'].apply(no_punc)

In [33]:
df_section.head(5)

tweetID                                               text  \
0  1072758722315862021  UK Prime Minister Theresa May will face a vote...   
1  1072959907958276096  UK PM Theresa May wins confidence vote with 20...   
2  1073329348755767303  The probe of the inaugural fund partly arises ...   
3  1073499538659393536  The week Brexit hit the brick wall :  Commons ...   
4  1073542862573371392  Have watched these kinds of pictures looking f...   

  topicName       usID             usName  edInput               crDate  \
0  Business    5402612  BBC Breaking News        2  2018-12-12 07:43:04   
1  Business    5402612  BBC Breaking News        2  2018-12-12 21:02:31   
2  Business  705706292   Rebecca Ballhaus        2  2018-12-13 21:30:32   
3  Business   25984418      Nick Robinson        1  2018-12-14 08:46:49   
4  Business   61183568   Laura Kuenssberg        2  2018-12-14 11:38:58   

                                              tokens  \
0  [uk, prime, minister, theresa, may, face, vote...   
1  [uk, pm, theresa, may, wins, confidence, vote,...   
2  [probe, inaugural, fund, partly, arises, mater...   
3  [week, brexit, hit, brick, wall, :, commons, ’...   
4  [watched, kinds, pictures, looking, body, lang...   

                                          tokens_url  \
0                          [https://t.co/wzta1v9scu]   
1  [https://t.co/9vw2goqodc, https://t.co/bafuusz...   
2                          [https://t.co/wvercqwnvg]   
3                                                 []   
4                          [https://t.co/1tg4h62z7l]   

                                          tokens_new  
0  [uk, prime, minister, theresa, may, face, vote...  
1  [uk, pm, theresa, may, wins, confidence, vote,...  
2  [probe, inaugural, fund, partly, arises, mater...  
3  [week, brexit, hit, brick, wall, commons, acce...  
4  [watched, kinds, pictures, looking, body, lang...

In [34]:
#Lemmatizing starts here

In [35]:
from nltk.stem import WordNetLemmatizer

In [36]:
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to /root/nltk_data...


True

In [37]:
nltk.download('omw-1.4')

[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


True

In [38]:
wordnet_lemmatizer = WordNetLemmatizer()

In [39]:
df_section['tokens_lemmatize'] = ''

In [40]:
def word_lemmatize(string):
  output = [wordnet_lemmatizer.lemmatize(w) for w in string]
  return output

df_section['tokens_lemmatize'] =df_section['tokens_new'].apply(word_lemmatize)

In [41]:
df_section.head(10)

tweetID                                               text  \
0  1072758722315862021  UK Prime Minister Theresa May will face a vote...   
1  1072959907958276096  UK PM Theresa May wins confidence vote with 20...   
2  1073329348755767303  The probe of the inaugural fund partly arises ...   
3  1073499538659393536  The week Brexit hit the brick wall :  Commons ...   
4  1073542862573371392  Have watched these kinds of pictures looking f...   
5  1074475798227226627  The world's most successful people start their...   
6  1074056878542471169  “This is Pearl Harbor. The whole country, and ...   
7  1074964829830557696  Jose Mourinho has been sacked by Manchester Un...   
8  1075269960086241281  Pursue your ideas, no matter how obvious they ...   
9  1075315807666794496  You don’t have to choose between being a leade...   

  topicName       usID              usName  edInput               crDate  \
0  Business    5402612   BBC Breaking News        2  2018-12-12 07:43:04   
1  Business    5402612   BBC Breaking News        2  2018-12-12 21:02:31   
2  Business  705706292    Rebecca Ballhaus        2  2018-12-13 21:30:32   
3  Business   25984418       Nick Robinson        1  2018-12-14 08:46:49   
4  Business   61183568    Laura Kuenssberg        2  2018-12-14 11:38:58   
5  Business   16896485                Inc.        2  2018-12-17 01:26:07   
6  Business   41154223  Saikat Chakrabarti        1  2018-12-15 21:41:29   
7  Business  265902729           BBC Sport        2  2018-12-18 09:49:21   
8  Business   14800270  Harvard Biz Review        1  2018-12-19 06:01:50   
9  Business   14800270  Harvard Biz Review        1  2018-12-19 09:04:01   

                                              tokens  \
0  [uk, prime, minister, theresa, may, face, vote...   
1  [uk, pm, theresa, may, wins, confidence, vote,...   
2  [probe, inaugural, fund, partly, arises, mater...   
3  [week, brexit, hit, brick, wall, :, commons, ’...   
4  [watched, kinds, pictures, looking, body, lang...   
5  [worlds, successful, people, start, day, 4, .,...   
6  [“, pearl, harbor, ., whole, country, ,, whole...   
7  [jose, mourinho, sacked, manchester, united, ....   
8  [pursue, ideas, ,, matter, obvious, seem, http...   
9  [’, choose, leader, life, https://t.co/sc9fi5v...   

                                          tokens_url  \
0                          [https://t.co/wzta1v9scu]   
1  [https://t.co/9vw2goqodc, https://t.co/bafuusz...   
2                          [https://t.co/wvercqwnvg]   
3                                                 []   
4                          [https://t.co/1tg4h62z7l]   
5                          [https://t.co/sef0u4xwp8]   
6                          [https://t.co/vposikgvim]   
7  [https://t.co/xl8mju4w1u, https://t.co/zdrd3vu...   
8                          [https://t.co/irqq39x9h2]   
9                          [https://t.co/sc9fi5v8zs]   

                                          tokens_new  \
0  [uk, prime, minister, theresa, may, face, vote...   
1  [uk, pm, theresa, may, wins, confidence, vote,...   
2  [probe, inaugural, fund, partly, arises, mater...   
3  [week, brexit, hit, brick, wall, commons, acce...   
4  [watched, kinds, pictures, looking, body, lang...   
5           [worlds, successful, people, start, day]   
6  [pearl, harbor, whole, country, whole, world, ...   
7  [jose, mourinho, sacked, manchester, united, f...   
8             [pursue, ideas, matter, obvious, seem]   
9                             [choose, leader, life]   

                                    tokens_lemmatize  
0  [uk, prime, minister, theresa, may, face, vote...  
1  [uk, pm, theresa, may, win, confidence, vote, ...  
2  [probe, inaugural, fund, partly, arises, mater...  
3  [week, brexit, hit, brick, wall, common, accep...  
4  [watched, kind, picture, looking, body, langua...  
5            [world, successful, people, start, day]  
6  [pearl, harbor, whole, country, whole, world, ...  
7  [jose, mourinho, sacked, manchester,

In [42]:
#usID and UsName has the same data

In [43]:
df_column_duplicate = df_section[df_section['usID'] == 2735591] 
df_column_duplicate['usName'].nunique()

0

In [44]:
# Filter out Topic Name Column and edInput -1,3,4
df_section = df_section[(df_section['edInput'] == 1) | (df_section['edInput'] == 2)]
#df_model_baseline = df_section[['usName','edInput','tokens']].copy()
df_model = df_section[['usName','edInput','tokens','tokens_lemmatize']].copy()

In [45]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics import confusion_matrix,accuracy_score,precision_score,recall_score,f1_score,classification_report
from sklearn.model_selection import cross_val_score
import time

In [46]:
#Create pipelines for categorical features
steps_cat_logistic = [('imputer', SimpleImputer(strategy ='most_frequent')),
                      ('onehot',OneHotEncoder(handle_unknown ='ignore'))]

steps_text_logistic = [('vectorizer', CountVectorizer())]

categorical_transformer = Pipeline(steps = steps_cat_logistic)
text_transformer = Pipeline(steps = steps_text_logistic)

categorical_features = ['usName']
text_features_baseline = 'tokens'
text_features_final = 'tokens_lemmatize'

In [47]:
preprocessor_logistic_baseline = ColumnTransformer(transformers = [('cat',categorical_transformer, categorical_features),
                                                          ('text',text_transformer,text_features_baseline)])

In [48]:
preprocessor_logistic = ColumnTransformer(transformers = [('cat',categorical_transformer, categorical_features),
                                                          ('text',text_transformer,text_features_final)])

In [49]:
#>>>>Logistic Regression<<<<<
#Trained Model with Both Baseline and Final Version.
#Predictions were similar on the test data
#Dimensions are same for both
#Training Final version was 1 second faster than baseline
#Final model slightly predicted better results than baseline. 

In [50]:
#Test Logistic Regression Model (Baseline)

In [51]:
logreg_base = Pipeline (steps = [('preprocessor_logistic',preprocessor_logistic_baseline),
                            ('regressor', LogisticRegression(max_iter=300))])

In [52]:
#Test Logistic Regression Model (Final)

In [53]:
logreg = Pipeline (steps = [('preprocessor_logistic',preprocessor_logistic),
                            ('regressor', LogisticRegression(max_iter=300))])

In [54]:
#df_model_baseline.info()

In [55]:
y = df_model.edInput 
X = df_model.drop('edInput', axis =1).astype(str)

In [56]:
X.head(20)

usName                                             tokens  \
0    BBC Breaking News  ['uk', 'prime', 'minister', 'theresa', 'may', ...   
1    BBC Breaking News  ['uk', 'pm', 'theresa', 'may', 'wins', 'confid...   
2     Rebecca Ballhaus  ['probe', 'inaugural', 'fund', 'partly', 'aris...   
3        Nick Robinson  ['week', 'brexit', 'hit', 'brick', 'wall', ':'...   
4     Laura Kuenssberg  ['watched', 'kinds', 'pictures', 'looking', 'b...   
5                 Inc.  ['worlds', 'successful', 'people', 'start', 'd...   
6   Saikat Chakrabarti  ['“', 'pearl', 'harbor', '.', 'whole', 'countr...   
7            BBC Sport  ['jose', 'mourinho', 'sacked', 'manchester', '...   
8   Harvard Biz Review  ['pursue', 'ideas', ',', 'matter', 'obvious', ...   
9   Harvard Biz Review  ['’', 'choose', 'leader', 'life', 'https://t.c...   
10    Business Insider  ['meet', 'brand', 'new', 'bugatti', 'divo', 'h...   
11        Tech Insider  ['add', 'second', 'monitor', 'laptop', 'https:...   
12        Tech Insider  ['utah', 'drops', 'fish', 'airplanes', 'mounta...   
13             INSIDER  ['spent', 'day', 'one', 'sanctuarys', 'keepers...   
14    Business Insider  ['', 'aquaman', '', 'already', 'made', 'money'...   
15             INSIDER  ['heres', '8', '.', '.', 'looks', 'like', 'aro...   
16    Bloomberg Crypto  ['.', '@novogratz', 'putting', 'money', 'mouth...   
17    Business Insider  ['exercise', 'scientist', 'explains', 'shouldn...   
18    Bloomberg Crypto  ['punishing', 'year', ',', 'crypto', 'enthusia...   
19    Business Insider  ['hondas', '$', '4.9', 'million', 'private', '...   

                                     tokens_lemmatize  
0   ['uk', 'prime', 'minister', 'theresa', 'may', ...  
1   ['uk', 'pm', 'theresa', 'may', 'win', 'confide...  
2   ['probe', 'inaugural', 'fund', 'partly', 'aris...  
3   ['week', 'brexit', 'hit', 'brick', 'wall', 'co...  
4   ['watched', 'kind', 'picture', 'looking', 'bod...  
5   ['world', 'successful', 'people', 'start', 'day']  
6   ['pearl', 'harbor', 'whole', 'country', 'whole...  
7   ['jose', 'mourinho', 'sacked', 'manchester', '...  
8     ['pursue', 'idea', 'matter', 'obvious', 'seem']  
9                        ['choose', 'leader', 'life']  
10        ['meet', 'brand', 'new', 'bugatti', 'divo']  
11             ['add', 'second', 'monitor', 'laptop']  
12  ['utah', 'drop', 'fish', 'airplane', 'mountain...  
13  ['spent', 'day', 'one', 'sanctuary', 'keeper',...  
14  ['aquaman', 'already', 'made', 'money', 'produ...  
15        ['here', 'look', 'like', 'around', 'world']  
16  ['putting', 'money', 'mouth', 'come', 'cryptoc...  
17  ['exercise', 'scientist', 'explains', 'shouldn...  
18  ['punishing', 'year', 'crypto', 'enthusiast', ...  
19  ['hondas', 'million', 'private', 'jet', 'absol...

In [57]:
X_train, X_test,y_train,y_test = train_test_split(X,y,test_size = 0.2,random_state =42)

In [58]:
t0 = time.time()
log_model_base = logreg_base.fit(X_train,y_train) #Train Model For Baseline
print("Training time Baseline:", time.time()-t0)

t1 = time.time()
log_model = logreg.fit(X_train,y_train) # Train Model After Clearing Text
print("Training time Final:", time.time()-t1)

Training time Baseline: 6.254120111465454
Training time Final: 2.3513870239257812


In [59]:
y_pred_base = log_model_base.predict(X_test)
y_pred = log_model.predict(X_test)

In [60]:
log_model.score(X_train, y_train)#Final score 

0.9286725516322452

In [61]:
log_model.score(X_test, y_test)#Final score 

0.8259783513738551

In [62]:
log_model_base.score(X_train, y_train) #Baseline score 

0.9679796802131913

In [63]:
log_model_base.score(X_test, y_test)#Baseline score

0.8239800166527893

In [64]:
confusion = confusion_matrix(y_test,y_pred)
confusion
#TP + TN  = 3075 + 1885 = 4960 (predicted right)
#FP + FN = 526 + 519 = 1045 (predicted wrong)

array([[3075,  526],
       [ 519, 1885]])

In [65]:
print(classification_report(y_test,y_pred))
#Precision:Accuracy of positive predictions. TP/(TP + FP)- not to label an instance positive that is actually negative
#Recall: Fraction of positives that were correctly identified. TP/(TP+FN)
#The F1 score is a weighted harmonic mean of precision and recall such that the best score is 1.0 and the worst is 0.0. (What percent of positive predictions were correct)
#F1 Score = 2*(Recall * Precision) / (Recall + Precision)

              precision    recall  f1-score   support

           1       0.86      0.85      0.85      3601
           2       0.78      0.78      0.78      2404

    accuracy                           0.83      6005
   macro avg       0.82      0.82      0.82      6005
weighted avg       0.83      0.83      0.83      6005



In [66]:
confusion = confusion_matrix(y_test,y_pred_base)
confusion
#TP + TN  = 3053 + 1895 = 4948 (predicted right)
#FP + FN = 548 + 509 = 1057 (predicted wrong)

array([[3053,  548],
       [ 509, 1895]])

In [67]:
print(classification_report(y_test,y_pred_base))

              precision    recall  f1-score   support

           1       0.86      0.85      0.85      3601
           2       0.78      0.79      0.78      2404

    accuracy                           0.82      6005
   macro avg       0.82      0.82      0.82      6005
weighted avg       0.82      0.82      0.82      6005



In [68]:
cv_results = cross_val_score(logreg,X_train,y_train, cv = 10)

In [69]:
print('Cross-Validation Accuracy Scores', cv_results) # Results are consistent - Average 0.82
cv_results.mean()

Cross-Validation Accuracy Scores [0.81140716 0.82056619 0.82597835 0.83180683 0.82431307 0.81182348
 0.818409   0.82548938 0.82382341 0.82340691]


0.8217023783803654

In [70]:
#>>>>Decision Tree Classifier<<<<<

In [71]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score


In [79]:
df_model.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 30021 entries, 0 to 30023
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   usName            30021 non-null  object
 1   edInput           30021 non-null  int64 
 2   tokens            30021 non-null  object
 3   tokens_lemmatize  30021 non-null  object
dtypes: int64(1), object(3)
memory usage: 1.1+ MB


In [132]:
#Create Pipelines for Categorical and Text Features

steps_cat_decision = [('imputer', SimpleImputer(strategy ='most_frequent')),
                      ('onehot',OneHotEncoder(handle_unknown = 'ignore'))]

steps_text_decision = [('vector', CountVectorizer())]


categorical_transformer = Pipeline(steps = steps_cat_decision)
text_transformer = Pipeline(steps = steps_text_decision)

cat_features = ['usName']
text_features_baseline = 'tokens'
text_features_final = 'tokens_lemmatize'


#ColumnTransformer
preprocessor_decision_baseline = ColumnTransformer(transformers = [('cat',categorical_transformer,cat_features),
                                                                   ('text', text_transformer,text_features_baseline)])

preprocessor_decision_final = ColumnTransformer(transformers = [('cat',categorical_transformer,cat_features),
                                                                ('text',text_transformer,text_features_final)])

#Pipeline
decision_tree_base = Pipeline(steps = [('preprocessor_decision_baseline', preprocessor_decision_baseline),
                                       ('decision_tree', DecisionTreeClassifier(max_depth = 9, random_state =42))])

decision_tree = Pipeline(steps = [('preprocessor_decision-final',preprocessor_decision_final),
                                  ('decision_tree', DecisionTreeClassifier(max_depth = 9, random_state =42))])


#Define X and y
y = df_model.edInput 
X = df_model.drop('edInput',axis =1).astype(str)


In [133]:
X_train, X_test,y_train,y_test = train_test_split(X,y,test_size = 0.2,random_state =42)

#Fit
decision_tree_model_base = decision_tree_base.fit(X_train,y_train)

decision_tree_model = decision_tree.fit(X_train,y_train)

#Predict
t0 = time.time()
y_pred_tree_baseline = decision_tree_model_base.predict(X_test)
print("Training time Baseline:", time.time()-t0)

t1 = time.time()
y_pred_tree_final = decision_tree_model.predict(X_test)
print("Training time Final:", time.time()-t1)

#Accuracy
acc_baseline = accuracy_score(y_test,y_pred_tree_baseline)
print('Accuracy Score - Baseline: '+ str(acc_baseline))

acc_final = accuracy_score(y_test,y_pred_tree_final)
print('Accuracy Score - Final: '+ str(acc_final))

#Cross Validation
cv_results_base = cross_val_score(decision_tree_base,X_train,y_train, cv=10)
print('CV Results - Baseline: ' + str(cv_results_base.mean()))

cv_results_final = cross_val_score(decision_tree, X_train,y_train, cv=10)
print('CV Results - Final: ' + str(cv_results_final.mean()))

Training time Baseline: 0.12532281875610352
Training time Final: 0.09058761596679688
Accuracy Score - Baseline: 0.6454621149042464
Accuracy Score - Final: 0.6144879267277269
CV Results - Baseline: 0.6566034447900385
CV Results - Final: 0.6187549005566304


In [122]:
import numpy as np
from sklearn.model_selection import GridSearchCV

In [124]:
#GridSearchCV
tree_param = {'decision_tree__max_depth':np.arange(1,10,2)}
#'criterion':['gini','entropy'],'max_depth':[4,5,6,7,8,9,10,11,12,15,20,30,40,50,70,90,120,150]

grid_search_base = GridSearchCV(decision_tree_base,tree_param,scoring='accuracy', cv =5)
grid_search_base.fit(X_train,y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('preprocessor_decision_baseline',
                                        ColumnTransformer(transformers=[('cat',
                                                                         Pipeline(steps=[('imputer',
                                                                                          SimpleImputer(strategy='most_frequent')),
                                                                                         ('onehot',
                                                                                          OneHotEncoder(handle_unknown='ignore'))]),
                                                                         ['usName']),
                                                                        ('text',
                                                                         Pipeline(steps=[('vector',
                                                                                          CountVectorizer())]),
                                                                         'tokens')])),
                                       ('decision_tree',
                                        DecisionTreeClassifier(max_depth=6,
                                                               random_state=42))]),
             param_grid={'decision_tree__max_depth': array([1, 3, 5, 7, 9])},
             scoring='accuracy')

In [127]:
print(grid_search_base.best_estimator_)

Pipeline(steps=[('preprocessor_decision_baseline',
                 ColumnTransformer(transformers=[('cat',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('onehot',
                                                                   OneHotEncoder(handle_unknown='ignore'))]),
                                                  ['usName']),
                                                 ('text',
                                                  Pipeline(steps=[('vector',
                                                                   CountVectorizer())]),
                                                  'tokens')])),
                ('decision_tree',
                 DecisionTreeClassifier(max_depth=9, random_state=42))])


In [130]:
#GridSearchCV
tree_param = {'decision_tree__max_depth':np.arange(1,10,2)}
#'criterion':['gini','entropy'],'max_depth':[4,5,6,7,8,9,10,11,12,15,20,30,40,50,70,90,120,150]

grid_search_final = GridSearchCV(decision_tree,tree_param,scoring='accuracy', cv =5)
grid_search_final.fit(X_train,y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('preprocessor_decision-final',
                                        ColumnTransformer(transformers=[('cat',
                                                                         Pipeline(steps=[('imputer',
                                                                                          SimpleImputer(strategy='most_frequent')),
                                                                                         ('onehot',
                                                                                          OneHotEncoder(handle_unknown='ignore'))]),
                                                                         ['usName']),
                                                                        ('text',
                                                                         Pipeline(steps=[('vector',
                                                                                          CountVectorizer())]),
                                                                         'tokens_lemmatize')])),
                                       ('decision_tree',
                                        DecisionTreeClassifier(max_depth=6,
                                                               random_state=42))]),
             param_grid={'decision_tree__max_depth': array([1, 3, 5, 7, 9])},
             scoring='accuracy')

In [131]:
print(grid_search_final.best_estimator_)

Pipeline(steps=[('preprocessor_decision-final',
                 ColumnTransformer(transformers=[('cat',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='most_frequent')),
                                                                  ('onehot',
                                                                   OneHotEncoder(handle_unknown='ignore'))]),
                                                  ['usName']),
                                                 ('text',
                                                  Pipeline(steps=[('vector',
                                                                   CountVectorizer())]),
                                                  'tokens_lemmatize')])),
                ('decision_tree',
                 DecisionTreeClassifier(max_depth=9, random_state=42))])


In [ ]:
df_section.info()

In [ ]:
#use this part to split the text column
#X = df_section.iloc[4:5,5].to_string().split(",")

In [ ]:
#Notes:
#    1. Check urllinks - related to business websites
#    2. Agirlik ortalamasi belirlemek 
#    3. drop duplicates